### ISSUE TITLE
Umbraco Forms 'Data' directory

### ISSUE THREAD

**Issue by snerpton** [ElijahGlover/Umbraco-S3-Provider#16]
Hi,

I've had great success using `Umbraco.Storage.S3` to offload Umbraco's media directory to Amazon's S3, but I'm wondering if it is possible to do the same for Umbraco Forms's data directory `/App_Data/UmbracoForms/Data`?

Current set up in case it is useful...

I have an S3 bucket with the following structure:

- Data
  - forms
    - guid-01.json
  - Templates
    -  form-01.json
    -  form-02.json
  - Workflows
    - workflow-01.json
    - workflow-02.json
    - workflow-03.json
    - workflow-04.json
    - workflow-05.json

I have added the following to `~/Config/FileSystemProviders.config`:

```
  <Provider alias="forms" type="Umbraco.Storage.S3.BucketFileSystem, Umbraco.Storage.S3">
    <Parameters>
      <add key="bucketName" value="my-bucket-name.com" />
      <add key="bucketHostName" value="" />
      <add key="bucketKeyPrefix" value="Data" />
      <add key="region" value="eu-west-2" />
    </Parameters>
  </Provider>  
```

and added `FileSystemVirtualPathProvider.Configure<FileSystemWrapper>("Data");` to an `ApplicationStarting` method I have created in `Global.asax.cs`.

**Comment by PhilA1** [comment]
I didn't do anything in global but I added the relevant config.
Creating and saving a form works fine.
Trying to see the forms in the tree just throws 404 errors in the console.

 <Provider alias="forms" type="Umbraco.Storage.S3.BucketFileSystem, Umbraco.Storage.S3">
    <Parameters>
      <add key="bucketName" value="my-bucket-name.com" />
      <add key="bucketHostName" value="" />
      <add key="bucketKeyPrefix" value="forms-data" />
      <add key="region" value="eu-west-2" />
    </Parameters>
  </Provider>

**Comment by AndyButland** [comment]
Just wondering if you could share what you've found so far @PhilA1 please?  Coincidentally I was looking into the same problem yesterday, then found your comment here.  For me I'm noticing calls to GetDirectories and GetFiles seem to be constantly called on a loop, with an incorrect path - `forms/forms` instead of `forms`.

Have tried adding some simple code to correct the path, e.g....

```
            if (path == "forms/forms")
            {
                path = "forms/";
            }
```

... which does then allow the tree to load, but still can't load a form and also continue to see these regular calls to the methods mentioned above.  So not got very far basically!  Sounds like maybe you have?

**Comment by PhilA1** [comment]
Hi,

I’m working on putting in a pull request for this.
I have found a few bugs around GetFiles and GetDirectories along with RemovePrefix which both these use, along with the looping issue.  I also have the getFiles filter param working.

I have them working in a test harness and I have just started testing in Umbraco.
Umbraco can now see my file and display its name in the tree but I’m seeing errors now that need to be fixed.

Regards

Phil

**Comment by AndyButland** [comment]
Nice one.  Well I'll follow any updates on this issue now, and hopefully can help test when you get the PR together.

**Comment by PhilA1** [comment]
I’m back on it today so we’ll see where I get to.

**Comment by PhilA1** [comment]
Almost there.
It creates forms using templates in S3, and it saves to S3.
Tree structure loads etc.
Edit/Delete may need a fix…..

[cid:image001.png@01D33136.DAC624D0]

**Comment by PhilA1** [comment]
Working, pull request is in: https://github.com/ElijahGlover/Umbraco-S3-Provider/pull/17

For the time being I will use my fork but I’ll jump back once it’s all merged and done.

**Comment by ElijahGlover** [comment]
Pull request #17 has been merged. It would be nice if someone could write unit tests around filtering.

**Comment by PhilA1** [comment]
Unit tests PR in

**Comment by PhilA1** [comment]
Can you let me know when the new nuget package is published please.

**Comment by ElijahGlover** [comment]
There was a problem with Appveyor, automatic builds should be enabled for next time. Released version is `1.0.25`

https://www.nuget.org/packages/Umbraco.Storage.S3/1.0.25

**Comment by PhilA1** [comment]
Something odd with that package, it wants to add loads of new dependencies that were not there before.
I’ve consumed your project into my solution for the time being and that hasn’t required these dependencies.
Very odd:
[cid:image001.png@01D332BB.B40F6FA0]
